lib/ui/painting.dart

```diff
@@ -1607,12 +1655,152 @@ class Path extends NativeFieldWrapperClass2 {
+/// would get two PathMetric objects in this iterable.  Call 
+/// [Path.computeMetrics] to create the iterable.
+class PathMetrics extends collection.IterableBase<PathMetric> {
+  final Iterator<PathMetric> _iterator;
+
+  PathMetrics._(Path path, bool forceClosed) :
+    _iterator = new PathMetricIterator._(new PathMetric._(path, forceClosed));
+
+  @override
+  Iterator<PathMetric> get iterator => _iterator;
+}
+
+/// Tracks iteration from one segment of a path to the next for measurement.
+class PathMetricIterator implements Iterator<PathMetric> {
+  PathMetric _pathMetric;
+  bool _firstTime = true;
+
+  PathMetricIterator._(this._pathMetric);
+
+  @override
+  PathMetric get current => _firstTime ? null : _pathMetric;
+
+  @override
+  bool moveNext() {
+    // PathMetric isn't a normal iterable - it's already initialized to its
+    // first Path.  Should only call _moveNext when done with the first one. 
+    if (_firstTime == true) {
+      _firstTime = false;
+      return true;
+    } else if (_pathMetric?._moveNext() == true) {
+      return true;
+    } else {
+      _pathMetric = null;
+      return false;
+    }
+  }
+}
+
+/// Utilities for measuring a [Path] and extracting subpaths.
+///
+/// Call [Path.computeMetrics] to create these objects. Once created, measures
+/// will only be valid while the path remains unmodified.  If the path is 
+/// modified, the behavior of the PathMetric object is undefined.
+class PathMetric extends NativeFieldWrapperClass2 {
+  /// Create a new empty [Path] object.
+  PathMetric._(Path path, bool forceClosed) { _constructor(path, forceClosed); }
+  void _constructor(Path path, bool forceClosed) native 'PathMeasure_constructor';
+
+  /// Return the total length of the current contour.
+  double get length native 'PathMeasure_getLength';
+
+  /// Pins distance to 0 <= distance <= getLength(), and then computes the 
+  /// corresponding position and tangent.
+  /// 
+  /// The [Tangent] object returns contains the position and angle (in radians)
+  /// of the tangent at specified distance. 
+  /// Returns null for a null or zero-length [Path].
+  Tangent getTangentForOffset(double distance) {
+    final Float32List posTan = _getPosTan(distance);
+    // first entry == 0 indicates that Skia returned false
+    if (posTan[0] == 0.0) {
+      return null;
+    } else {
+      return new Tangent(
+        new Offset(posTan[1], posTan[2]), 
+        math.atan2(posTan[4], posTan[3])
+      );
+    }
+  }
+  Float32List _getPosTan(double distance) native 'PathMeasure_getPosTan';
+
+  /// Given a start and stop distance, return the intervening segment(s).
+  /// 
+  /// startD and stopD are pinned to legal values (0..[length])
+  /// Returns null if the segment is 0 length or startD > stopD.
+  /// Begin the segment with a moveTo if startWithMoveTo is true.
+  Path extractPath(double startD, double endD, bool startWithMoveTo) native 'PathMeasure_getSegment';
+
+  /// Returns true if the path is closed.
```

Comment:
```
/// Whether the contour is closed.
///
/// Returns true if the contour ends with a call to [Path.close] (which may have been implied, e.g.
/// when using [Path.addRect]). Returns false otherwise.
```

Add a test that verifies that Path.addRect creates a contour that returns true here.